// g1/rollup.config.js
import uglify from 'rollup-plugin-uglify'
import htmlparts from './rollup-plugin-htmlparts.js'
import resolve from 'rollup-plugin-node-resolve'
import commonjs from 'rollup-plugin-commonjs'
import babel from 'rollup-plugin-babel'
import babelrc from 'babelrc-rollup'

const babelConfig = {
  'presets': ['env']
}

export default [
  {
    input: "index",
    output: {
      file: "dist/g1.js",
      format: "umd",
      name: "g1",
      extend: true
    },
    plugins: [
      resolve(),
      commonjs(),
      htmlparts('src/formhandler.template.html'),
      htmlparts('src/dropdown.template.html')
    ]
  },
  {
    input: "index",
    output: {
      file: "dist/g1.min.js",
      format: "umd",
      name: "g1",
      extend: true
    },
    plugins: [
      resolve(),
      commonjs(),
      htmlparts('src/formhandler.template.html'),
      htmlparts('src/dropdown.template.html'),
      uglify()
    ]
  },
  {
    input: "index-datafilter",
    output: { file: "dist/datafilter.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-urlfilter",
    output: { file: "dist/urlfilter.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-formhandler",
    output: { file: "dist/formhandler.min.js", format: "umd", name: "g1", extend: true},
    plugins: [
      htmlparts('src/formhandler.template.html'),
      process.env.npm_lifecycle_event == 'dev' ? '' : uglify()
    ],
  },
  {
    input: "index-highlight",
    output: { file: "dist/highlight.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-template",
    output: { file: "dist/template.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-translate",
    output: { file: "dist/translate.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-event",
    output: { file: "dist/event.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-leaflet",
    output: { file: "dist/leaflet.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-mapviewer",
    output: {
      file: "dist/mapviewer.min.js", format: "umd", name: "g1", extend: true, globals: {
        leaflet: 'L',
        d3: 'd3'
      }
    },
    plugins: [
      resolve(),
      commonjs(),
      babel(babelrc({ config: babelConfig, exclude: 'node_modules/**' })),
      process.env.npm_lifecycle_event == 'dev' ? '' : uglify()
    ],
    // indicate which modules should be treated as external
    external: ['leaflet', 'd3']
  },
  {
    input: "index-sanddance",
    output: { file: "dist/sanddance.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-scale",
    output: { file: "dist/scale.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()],
  },
  {
    input: "index-urlchange",
    output: { file: "dist/urlchange.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-ajax",
    output: { file: "dist/ajax.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-dropdown",
    plugins: [
      resolve(),
      commonjs(),
      htmlparts('src/dropdown.template.html')
    ],
    output: {
      file: "dist/dropdown.min.js",
      format: "umd",
      name: "g1",
      extend: true
    }
  },
  {
    input: "index-search",
    plugins: [
      resolve(),
      commonjs()
    ],
    output: {
      file: "dist/search.min.js",
      format: "umd",
      name: "g1",
      extend: true
    }
  },
  {
    input: "index-fuzzysearch",
    output: { file: "dist/fuzzysearch.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  },
  {
    input: "index-types",
    output: { file: "dist/types.min.js", format: "umd", name: "g1", extend: true },
    plugins: [uglify()]
  }
]
